file-attribute-caching/src/main/kotlin/com/pkware/file/fileAttributeCaching/FileAttributeCachingPath.kt

```diff
@@ -0,0 +1,209 @@
+ */
+internal class FileAttributeCachingPath(
+    private val fileSystem: FileSystem,
+    internal val delegate: Path
+) : ForwardingPath(delegate) {
+
+    // ExpirableCache for thisFileAttributeCachingPath a BasicFileAttributes cache
+    private val attributeCache = ExpirableCache<String, BasicFileAttributes?>(
+        TimeUnit.SECONDS.toMillis(CACHE_PRESERVATION_SECONDS)
+    )
+
+    override fun getFileSystem(): FileSystem = fileSystem
+
+    override fun resolve(other: Path): Path {
+        // Make sure if other is a FileAttributeCachingPath that we pass along other's
+        // delegate rather than other itself.
+        if (other is FileAttributeCachingPath) return delegate.resolve(other.delegate)
+        return delegate.resolve(other)
+    }
+
+    /**
+     * Sets the cache entry for the given attribute [name] with the given [value]. Can only set entire
+     * attribute `Class`es such as "dos:*", "posix:*", and "basic:*"
+     *
+     * The attribute name must include a "*" in order to be set within the cache.
+     *
+     * @param name The name of the attribute to cache.
+     * @param value The attribute value to cache.
+     */
+    fun <A : BasicFileAttributes?> setAttributeByName(name: String, value: A?) {
+        // remove basic from our attribute name if present as basicFileAttributes can be accessed without that qualifier
+        val checkedName = name.substringAfter("basic:")
+
+        // this check is to ensure that we are only storing attribute classes and not specific attributes
+        if (checkedName.contains("*")) {
+            attributeCache[checkedName] = value
+        }
+    }
+
+    /**
+     * Sets the cache entry for the given attribute `Class` [type] with the given [value].
+     *
+     * @param type The attribute `Class` to cache. `Class` types include [BasicFileAttributes], [DosFileAttributes],
+     * or [PosixFileAttributes].
+     * @param value The attribute value to cache.
+     */
+    fun <A : BasicFileAttributes?> setAttributeByType(type: Class<A>, value: A?) {
+        when (type) {
+            BasicFileAttributes::class.java -> attributeCache["*"] = value
+            DosFileAttributes::class.java -> attributeCache["dos:*"] = value
+            PosixFileAttributes::class.java -> attributeCache["posix:*"] = value
+        }
+    }
+
+    /**
+     * Copies this [FileAttributeCachingPath]s cached values to the [target]. Also runs [functionToExecute] midway
+     * through function execution to provide support to external operations such as
+     * [FileAttributeCachingFileSystemProvider.copy] and [FileAttributeCachingFileSystemProvider.move].
+     *
+     * @param target The [FileAttributeCachingPath] to copy cached attributes to.
+     * @param functionToExecute The [Runnable] function to execute midway through this function.
+     */
+    fun copyCachedAttributesTo(target: FileAttributeCachingPath, functionToExecute: Runnable) {
```

Comment:
I'm not seeing this used anywhere. Definitely worth adding a test to provide guidance on recommended use and bolster testing coverage.

Reply:
it's used in the FileAttributeCachingFileSystemProvider.kt copy and move functions which do have tests- that should be enough no?